src/Text/Layout/Table/Cell.hs

```diff
@@ -10,68 +13,169 @@ import Text.Layout.Table.Spec.OccSpec
+        Just matchRemaining | matchAt < -l || matchRemaining < -r -> AlignInfo (max 0 $ matchAt + matchRemaining + 1 + l + r) Nothing
+        -- There is a match, and it is not cut off
+        Just matchRemaining -> AlignInfo (matchAt + l) (Just $ matchRemaining + r)
+      where
+        AlignInfo matchAt mMatchRemaining = measureAlignment f a
+    buildCellView = buildCellView . join
+    buildCell = buildCellView
 
 instance (Cell a, Cell b) => Cell (Either a b) where
-    dropLeft n = bimap (dropLeft n) (dropLeft n)
-    dropRight n = bimap (dropRight n) (dropRight n)
-    dropBoth l r = bimap (dropBoth l r) (dropBoth l r)
     visibleLength = either visibleLength visibleLength
     measureAlignment p = either (measureAlignment p) (measureAlignment p)
+    buildCellView (CellView a l r) = either go go a
+      where
+        go x = buildCellView $ CellView x l r
     buildCell = either buildCell buildCell
 
 instance Cell String where
-    dropLeft = drop
-    dropRight n s = zipWith const s (drop n s)
     visibleLength = length
     measureAlignment p xs = case break p xs of
         (ls, rs) -> AlignInfo (length ls) $ case rs of
             []      -> Nothing
             _ : rs' -> Just $ length rs'
 
+    buildCellView = buildCellViewLRHelper stringB drop (\n s -> zipWith const s $ drop n s)
     buildCell = stringB
 
 instance Cell T.Text where
-    dropLeft = T.drop
-    dropRight = T.dropEnd
     visibleLength = T.length
     measureAlignment p xs = case T.break p xs of
         (ls, rs) -> AlignInfo (T.length ls) $ if T.null rs
             then Nothing
             else Just $ T.length rs - 1
 
+    buildCellView = buildCellViewLRHelper textB T.drop T.dropEnd
     buildCell = textB
 
+-- | Construct 'buildCellView' from a builder function, a function for
+-- trimming from the left, and a function for trimming from the right.
+--
+-- Used to define instances of 'Cell'.
+buildCellViewLRHelper :: StringBuilder b
+                      => (a -> b)  -- ^ Builder function for 'a'.
+                      -> (Int -> a -> a)  -- ^ Function for trimming on the left.
+                      -> (Int -> a -> a)  -- ^ Function for trimming on the right.
+                      -> CellView a
+                      -> b
+buildCellViewLRHelper build trimL trimR =
+    buildCellViewHelper build build build trimL trimR (\l r -> trimL l . trimR r)
+
+-- | Construct 'buildCellView' from a builder function, and a function for
+-- trimming from the left and right simultaneously.
+--
+-- Used to define instanced of 'Cell'.
+buildCellViewBothHelper :: StringBuilder b
+                        => (a -> b)  -- ^ Builder function for 'a'.
+                        -> (Int -> Int -> a -> a)  -- ^ Function for trimming on the left and right simultaneously.
+                        -> CellView a
+                        -> b
+buildCellViewBothHelper build trimBoth =
+    buildCellViewHelper build build build (`trimBoth` 0) (trimBoth 0) trimBoth
```

Comment:
Suggestion (but not requirement, feel free to resolve):  `Data.Function.flip` seems more readable to me.